### DISCUSSION TITLE
[Outdated] Meta: how should we best use the Discussions feature

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#4] [General]
E.g., regarding bechmarks: should we have one open discussion for each family of bechmarks we discuss or all in one discussion? Should we format benchmarks in any specific way, provide any metadata, particular kind of links to code to full results in CI, etc., or is informal description the best? How many results to include (e.g., in case of many sizes of datasets or many variants of the code)?

**Comment by Mikolaj** [comment]
Also, are html reports like the following one useful (it's committed to the repo, but can be viewed only via an external tool)?

https://combinatronics.com/Mikolaj/mostly-harmless/b9f1ed13b7e549f7f221f2ea4a1033ce59677247/misc-bench/ProdAll-78ba46b.html

Edit: and here is the report for a MNIST benchmark:

https://combinatronics.com/Mikolaj/mostly-harmless/1404371c24a0f173cacebe4e7ce747d5c7bda260/misc-bench/MnistAll-f24cffc.html

The `manual` run corresponds to the following manually written hmatrix gradient expression:

https://github.com/Mikolaj/mostly-harmless/blob/5a666216a11c79784a1444befa901751c54d1213/backprop-bench/MnistBackpropTools.hs#L388-L420

The other backprop runs probably compute a similar gradient, derived from a similar number of matrix operations and their handwritten gradients. That's probably the secret of the stellar performance of backprop in this benchmark (both flavours of backprop: bp-lens and bp-hkd).

Library ad couldn't keep up training the largest neural nest of 472660 parameters, but results for smaller nets are included. I didn't attempt writing a backprop instance computing gradient from each scalar addition and multiplication, because in the much simpler `prod` benchmark (gradients provided only for scalar operations, just as here for our library and for ad) its runtime was too long to be measured already at 10000 parameters.

As with the `prod` benchmark, this is not fully tested. There are reasonable tests for our implementation, but I haven't replicated the extensive tests for backprop-provided implementations and the implementation for library ad was not tested even a single time.

Edit2: I've since identified and fixed a memory leak in my MNIST implementation using library ad (as usual, slightly miscalculated number of parameters causing accumulation of thunks due to boxed vectors in ad). However, while the fix eliminates any leaks persisting from one gradient computation to another and so improves residency, it doesn't improve runtime. The problem may be that the big MNIST data is being translated to a boxed representation for the use by library ad, increasing heap size and slowing down GC. This can be mitigated to an extent by translating only the data that is known to be used in a given benchmark, but this would complicate code that is already complicated by the need to ensure the translation happens outside the benchmark and is fully forced, not to skew benchmark results.

Edit3: I now crop MNIST samples pre-emptively in benchmarks to avoid penalizing libraries that have to convert them and that would store the whole (huge especially after the conversion) set in memory slowing down GC. Alas, it didn't help with runtime, either.

Edi4: I now force parameters after each gradient computation with ad; 15% speedup, still lots of thunks, just changed the moment they are evaluated. No change to the big picture.
